### ISSUE TITLE
Add simple HTTP/2 extension framework

### ISSUE THREAD

**Issue by tatsuhiro-t** [nghttp2/nghttp2#385]
https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-5.5 describes the possibility of extensions in HTTP/2.  In this ticket, we'd like to add HTTP/2 extension framework to libnghttp2 which allows applications to submit and receive "simple" extension frames.  We called "simple" because these frames must not change the semantics of existing protocol components.

We already have a way to express extension frame in our current API.  That is `nghttp2_extension` struct.  In this ticket, we will add the APIs which can allow for the applications to perform following things:

* Send extension frames with arbitrary frame type, flags and payload.  Encoding frame payload will be delegated to callback function applications provide.
* Receive extension frames with arbitrary frame type and payload.  Decoding frame payload will be delegated to callback function applications provide.  The payload data will be passed to application through callback in chunk of data just like we do when receiving DATA frame.
* For the first iteration, we will limit the length of payload to 16384.  This is reasonable restriction, since larger frame length will destroy multiplexing and prioritization.

**Comment by tatsuhiro-t** [comment]
Here is proposed new APIs to achieve the goal:

Receiver side:

```c
/**
 * @functypedef
 *
 * Callback function invoked when chunk of extension frame payload is
 * received.  The |hd| points to frame header.  The received
 * chunk is |data| of length |len|.
 *
 * The implementation of this function must return 0 if it succeeds.
 *
 * To abort processing this extension frame, return
 * :enum:`NGHTTP2_ERR_CANCEL`.
 *
 * If fatal error occurred, application should return
 * :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.  In this case,
 * `nghttp2_session_recv()` and `nghttp2_session_mem_recv()` functions
 * immediately return :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.  If the
 * other values are returned, currently they are treated as
 * :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.
 */
typedef int (*nghttp2_on_extension_chunk_recv_callback)(
    nghttp2_session *session, const nghttp2_frame_hd *hd, const uint8_t *data,
    size_t len, void *user_data);

/**
 * @functypedef
 *
 * Callback function invoked when library asks the application to
 * unpack extension payload from its wire format.  The extension
 * payload has been passed to the application using
 * :type:`nghttp2_on_extension_chunk_recv_callback`.  The frame header
 * is already unpacked by the library and provided as |hd|.
 *
 * The implementation of this function stores the created object as a
 * result of unpacking in |*payload|, and returns 0.  The pointer
 * stored in |*payload| is opaque to the library, and the library does
 * not own its pointer.
 *
 * To abort processing this extension frame, return
 * :enum:`NGHTTP2_ERR_CANCEL`.
 *
 * If fatal error occurred, application should return
 * :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.  In this case,
 * `nghttp2_session_recv()` and `nghttp2_session_mem_recv()` functions
 * immediately return :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.  If the
 * other values are returned, currently they are treated as
 * :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.
 */
typedef int (*nghttp2_unpack_extension_callback)(nghttp2_session *session,
                                                 void **payload,
                                                 const nghttp2_frame_hd *hd,
                                                 void *user_data);

```

Sender side:
```c
/**
 * @functypedef
 *
 * Callbck function invoked when library asks the application to pack
 * extension payload to its wire format.  The frame header will be
 * packed by library.  Application must pack payload only.
 * frame->ext.payload is the object passed to
 * `nghttp2_submit_extension()` as payload parameter.  The |*flags| is
 * initialized to the flags parameter passed in
 * `nghttp2_submit_extension()`.  Application can modify flags value
 * if it wants.  Application must pack extension payload to the |buf|
 * of its capacity |len| bytes.
 *
 * The implementation of this function returns the number of bytes
 * written into |buf| when it succeeds.
 *
 * To abort processing this extension frame, return
 * :enum:`NGHTTP2_ERR_CANCEL`, and
 * :type:`nghttp2_on_frame_not_send_callback()` will be invoked.
 *
 * If fatal error occurred, application should return
 * :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.  In this case,
 * `nghttp2_session_send()` and `nghttp2_session_mem_send()` functions
 * immediately return :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.  If the
 * other values are returned, currently they are treated as
 * :enum:`NGHTTP2_ERR_CALLBACK_FAILURE`.
 */
typedef ssize_t (*nghttp2_pack_extension_callback)(nghttp2_session *session,
                                                   uint8_t *flags, uint8_t *buf,
                                                   size_t len,
                                                   const nghttp2_frame *frame,
                                                   void *user_data);

/**
 * @function
 *
 * Submits extension frame.
 *
 * Application can pass arbitrary frame flags and stream ID to |flags|
 * and |stream_id| respectively.  The |payload| is opaque pointer, and
 * it can be accessible though frame->ext.payload in
 * :type:`nghttp2_pack_extension_callback`.  The library will not own
 * passed |payload| pointer.
 *
 * This function returns 0 if it succeeds, or one of the following
 * negative error codes:
 *
 * :enum:`NGHTTP2_ERR_INVALID_STATE`
 *     If :type:`nghttp2_pack_extension_callback` is not set.
 * :enum:`NGHTTP2_ERR_NOMEM`
 *     Out of memory
 */
NGHTTP2_EXTERN int nghttp2_submit_extension(nghttp2_session *session,
                                            uint8_t type, uint8_t flags,
                                            int32_t stream_id, void *payload);
```

Here is the overview of how these new callbacks and function are used.

To receive extension frames, application sets `nghttp2_unpack_extension_callback` and `nghttp2_on_extension_chunk_recv_callback`.
When these callbacks are set, application is notified the beginning of extension frame by `nghttp2_on_begin_frame_callback`.  The payload of extension frame will be passed to the application by `nghttp2_on_extension_chunk_recv_callback`.  After whole frame payload was received, `nghttp2_unpack_extension_callback` is called.  Here, application is expected to unpack data from wire format, but it can be done while receiving chunk.  Finally, `nghttp2_on_frame_recv_callback` is called as usual.  The unpacked payload data is accessible via `frame->ext.payload`.

To send extension frames, application sets `nghttp2_pack_extension_callback`, and calls `nghttp2_submit_extension`.  The |payload| parameter can be arbitrary pointer, and opaque to the library.  This could be a pointer to user define struct, or already serialized frame payload.  Just before sending this frame, `nghttp2_pack_extension_callback` is called.  The `frame->ext.payload` will point to the opaque pointer passed to `nghttp2_submit_extension`.  The application is expected to pack extension payload to wire format.  Then `nghttp2_before_frame_send_callback` will be called if it set, and `nghttp2_on_frame_send_callback` will be called as usual.

**Comment by tatsuhiro-t** [comment]
#386 implements the proposed APIs.

**Comment by LPardue** [comment]
This looks very interesting.

Would `nghttp2_unpack_extension_callback` be called whenever *any* extension frame is received? I know extension frames are not in much wide usage but it might be annoying to be notified of an extension frame the receiver doesn't care about. These can be safely ignored according to the spec. 

So I wonder if the API could be designed to allow the receiver to filter in frame types of interest? Or alternatively, signal in the `on_begin_frame_callback` to not bother wasting time with `nghttp2_on_extension_chunk_recv_callback` or `nghttp2_unpack_extension_callback` for this stream.

**Comment by tatsuhiro-t** [comment]
Thank you for the comment.
It is possible to allow new error code (NGHTTP2_ERR_CANCEL) from `on_begin_frame_callback` for extension frame type to ignore rest of the callback.  It is the same mechanism we are doing in on_extension_chunk_recv_callback and unpack_extension_callback. But we don't want this return code for standard frames, since it could complicate exiting codebase.
I think just using switch-case in these new 2 callbacks and filtering out unwanted frame type by returning NGHTTP2_ERR_CANCEL also works fine, and probably not much overhead.  Also we probably do this anyway, to handle multiple extension frame type.

**Comment by tatsuhiro-t** [comment]
Sorry I did not fully read your comment.  I'm interesting in your idea to filter in frame types of interest.
Perhaps, we can have new function to declare user-defined frame types to receive.
This could work well, when we add official support of new frame types --- existing user code registers their own handling.  If it does, then we don't call our new frame handling code, and just let run user defined one.

**Comment by LPardue** [comment]
I think both of these strategies would be really useful.

**Comment by tatsuhiro-t** [comment]
061a557 implements the idea in my previous comment.

**Comment by tatsuhiro-t** [comment]
Updated https://github.com/tatsuhiro-t/nghttp2/issues/385#issuecomment-147089422 to reflect the recent commit.

**Comment by LPardue** [comment]
Cool, only have time to skim the change. Does this allow for listening for multiple extension frame types?

**Comment by tatsuhiro-t** [comment]
Yes, call `nghttp2_option_set_user_recv_extension_type()` repeatedly with different types you want.

**Comment by tatsuhiro-t** [comment]
Still same callback cannot be set multiple times in one session, so we have to check frame type.

**Comment by LPardue** [comment]
> Still same callback cannot be set multiple times in one session, so we have to check frame type.

That seems fair enough, the user can just switch on frame types they registered and select correct processing.

**Comment by tatsuhiro-t** [comment]
Fixed with #386
